### DISCUSSION TITLE
How to generate a list where an element value depend on the previous element?

### DISCUSSION THREAD

**Discussion by dniel** [instancio/instancio#843] [Q&A]
Sample or doc how to, if possible, to generate a list of values where next item in list i dependent on the previous value.
I'm trying to generate a list of values, each new list item is dependent in the previous item generated. 

Is it possible?

**Comment by armandino** [comment]
@dniel could you provide an example of such a list?

**Comment by dniel** [comment]
Something like this itinerary where the journey in composed by multiple legs.
```
data class Itinerary(val legs: List<ItineraryLeg>)
data class ItineraryLeg(val arrival: DateTime, val departure: DateTime, val location: String)
```

I would like to generate a Itinerary containing a generated set of ItineraryLegs, where the number of ItineraryLeg items could be random but the arrival and departure time of each ItineraryLeg should be after the previous.

**Reply by armandino** [reply]
Thanks for the clarification. This should be doable with 2 steps. First, we generate each departure 1 day after the previous one. Second, we can set each the arrival date after the departure - e.g. by adding 3 hours - for each `ItineraryLeg` instance.

Here's an example (I replaced Kotlin data classes with java records):

```java
record Itinerary(List<ItineraryLeg> legs) {}
record ItineraryLeg(LocalDateTime arrival, LocalDateTime departure, String location) {}

@Test
void itinerary() {
    AtomicReference<LocalDateTime> dateHolder = new AtomicReference<>(LocalDateTime.now());

    Itinerary itinerary = Instancio.of(Itinerary.class)
            .supply(field(ItineraryLeg::departure), () -> dateHolder.getAndUpdate(date -> date.plusDays(1)))
            .assign(Assign.valueOf(ItineraryLeg::departure)
                    .to(ItineraryLeg::arrival)
                    .as((LocalDateTime departure) -> departure.plusHours(3))) // arrival is 3h after the departure
            .create();

    assertThat(itinerary.legs())
            .isNotEmpty()
            .allMatch(leg -> leg.departure().isBefore(leg.arrival()))
            .extracting(ItineraryLeg::departure)
            .isSorted();
}
```

You could also randomise the number of days between departures (instead of having them 1 day apart) by using a `Generator` instead of a `Supplier`:

```java
.supply(field(ItineraryLeg::departure), random -> dateHolder.getAndUpdate(date -> date.plusDays(random.intRange(1, 7))))
```

**Edit**: see #856 for an alternative approach using a custom generator.
